### DISCUSSION TITLE
3차 스프린트 사용자 시나리오

### DISCUSSION THREAD

**Discussion by kyum-q** [woowacourse-teams/2024-code-zap#354] [기획 관련]
# 페르소나

|     | 만두 (개발자 취업준비생) | 
|-----|---------------------------|
| **생년** | 98년생 | 
| **경력** | 예비개발자 | 
| **전공** | 비전공 | 
| **사용 환경** | Window 사용 | 
| **학년** | 5학년 1학기 | 
| **현재 놓인**<br>**상황** | - 우테코 최종 코딩테스트를 앞두고 있음 <br> - 자주 사용되는 코드를 관리해두지 않아 코드를 찾을 때 시간이 오래걸린다. <br> - 같은 유효성 검사 코드에서 삽질하는 기능이 있다. <br> - 자바의 정석 1회독 |
| **Needs** | - 빠르게 찾고 싶은 코드 템플릿이 있다. <br> - 자주 쓰는 코드 템플릿 미리 준비하고 싶다. |
| **Customer Job** | 최종 코딩테스트 7기 합격!!! |

<br>

# 시나리오 

## 🥟 만두 (예비 개발자 사용자)

**목표:** 우테코 최종 코딩테스트에 합격하기 위해 코드 스니펫을 빠르게 찾고 자주 쓰는 코드 템플릿을 준비하기

**시나리오:**

- 만두는 우테코 코딩테스트를 준비하기 위해 '코드잽'을 실행합니다.
- 코딩 테스트에서 자주 사용될 템플릿을 미리 작성합니다.
- 검색 기능을 사용해 필요한 코드 스니펫을 신속하게 찾아 삽입합니다.
- 코드를 사용하고 싶을 때 한번의 클릭으로 쉽게 코드를 재사용합니다.
- 만두는 원하는 기능의 코드를 빠르게 찾아 시간을 대폭 단축하여 우테코 최종 코딩테스트에 합격합니다.

**Comment by kyum-q** [comment]
## 페르소나 - 만두

**Comment by kyum-q** [comment]
# 1. 로그인/회원가입

> 페르소나인 만두를 통해 나온 기능이라고 하긴 어렵다
> - 만두가 혼자 사용할 서비스라면 로그인이 추가될 필요 없기 때문이다.

> 우리 프로젝트 목표에 따라 나온 기능에 가깝다
> - 팀의 스프린트 목표: 이번 스프린트가 끝나고 사용자를 유입하고 싶다.
> - 우리 서비스는 결과적으로 만두만 사용하는 서비스가 아닌 여러 명이 사용하는 서비스이다.


> 사용자 구분이 필요하다.  -> 로그인 기능을 추가한다.

**Reply by kyum-q** [reply]
> 팀의 스프린트 목표: 이번 스프린트가 끝나고 사용자를 유입하고 싶다.

### 우리 프로젝트의 주 목적
- 개인의 코드 템플릿을 관리하는 것

### 사용자가 유입되면 내 템플릿 관리를 위해 사용자 구분이 되어야한다.
- 회원 분리가 되지 않음 -> 내가 만든 템플릿 외에 다양한 템플릿이 추가되므로 관리의 편리성이 없어진다
- 회원 분리가 됨 -> 우리 프로젝트는 사용자가 한 명이다. 단 한 명만 템플릿 관리가 간편하다

### 결론
욕구
- 우리 프로젝트는 사용자가 여러명이고 싶다.
- 모든 유저들의 템플릿 관리가 간편하다.

결론
- 회원 분리를 통해 여러명의 사용자가 개인의 템플릿 관리를 간편하게 할 수 있다.

**Reply by kyum-q** [reply]
## [2차 스프린트 후 이야기했던 것] 인증/인가 지금 필요한가?
몰리: 인증 인가를 미리 고려하는게 나을 것 같다.
초롱: 그럼 우리는 내 템플릿을 먼저 보는건가?
제우스: 필요성을 먼저 찾자. 기능 단위로 뭘 할지 하지 말고. 서비스의 방향성을 바꾸더라도 해결해야한다.
몰리: 나중에 쓸 것 같은 코드를 저장하는
짱수: 우리가 꼭 억지로 쓰는 것일까? 일단 한번 써보는 식으로 서비스 유입을.

**Reply by kyum-q** [reply]
## 기획 회의 시

기능에 대한 우선순위를 작성할 때는 로그인은 중요한 기능은 아니었다.
하지만 시간 안에 가능하다고 판단했고, 4차 스프린트 때 사용자 유입을 하고자 로그인 기능을 구현하기로 했다.

![image](https://github.com/user-attachments/assets/0540c3dc-0d85-4ef7-89ac-2f29a8d020cd)

**Reply by kyum-q** [reply]
## 템플릿 구분 방법

1. 회원가입/로그인
2. 템플릿에 비밀번호

### 회원가입을 선택한 이유
우리의 목표는 (페르소나의 시나리오를 보면) 자신이 만든 코드템플릿을 관리 및 재사용을 편리하게 해주는 것이다.
즉 내가 만든 템플릿을 모아볼 수 있어야한다. 하지만 템플릿에 비밀번호를 주는 것은 로그인 기능에 비해 내가 만든 템플릿을 모아보기 어려워진다. 
> 굳이 템플릿에 비밀번호를 주는 식으로 하고 모아볼 수 있게 한다면? 
> 동일한 비밀번호를 쓴 템플릿들을 같은 사용자라고 봐야할까? 구분이 어렵다.

**Reply by zangsu** [reply]
이 글의 첫 번째 댓글이 회원 도메인이 추가되고 인증/인가가 생기게 된 가장 큰 이유인 것 같네요! 👍

**Comment by kyum-q** [comment]
# 2. 카테고리

**Reply by HoeSeong123** [reply]
- 등록된 스니펫이 많아지면 원하는 스니펫을 바로 찾기 힘들다.
- 검색만을 통해서도 찾기 쉬울 수 있지만, 과거 내가 어떤 제목으로 저장했는지 기억하기 힘들 수 있음

**Reply by kyum-q** [reply]
## 카테고리 (한 글에 하나만 있는 것)
글마다 카테고리를 두어 카테고리별로 템플릿을 조회할 수 있도록 하자

### 장점
- 카테고리로 1차 분류 한다.
- 내가 저장한 템플릿을 쉽게 파악할 수 있다.
  - [ex. 자바, 스프링, 자바스크립트, CSS] - 언어 분류 -> 현재 필요한 코드의 언어는 자바일 경우 쉽게 찾을 수 있다
  - [ex. 레벨1, 레벨2, 프로젝트1, 프로젝트2] - 언제 사용된 코드인지 분류 -> 프로젝트 1 때 구현한 코드 템플릿을 쉽게 찾을 수 있다

**Reply by HoeSeong123** [reply]
코드잽을 너무 잘 사용하고 있습니다.
트러블 슈팅을 하기 위해 코드를 남길 때도 사용을 하고 있고, 알고리즘 문제를 풀 때마다 자주 쓰이는 로직 등을 저장하고 있습니다.
어느덧 100개가 넘는 템플릿이 생겨버렸습니다.

오늘은 DFS 알고리즘 문제를 푸는 날.
어랏 이 비슷한 문제 저번에 풀었던 거 같은데? 어떤 로직 사용했었더라?
(코드잽을 뒤적거린다)
아 그냥 시간순으로만 정렬되어 있으니까 도통 찾기가 힘드네...
알고리즘 관련 템플릿들만 모아놓을 수는 없으려나...

**Reply by HoeSeong123** [reply]
(짱수)

만두는 이제 템플릿을 하나씩 저장하기 시작했다.
그런데 템플릿이 개수가 쌓일 수록, 나에게 필요한 템플릿을 찾는 것이 힘들어 졌다.
내가 필요한 템플릿을 빠르게 찾기 위한 기능이 필요하다고 느껴졌다.
템플릿을 분류하기 위한 카테고리 기능을 추가하였고, 템플릿이 포함하고 있는 자잘한 정보를 태그에 담기로 하였다.
또한, 지금 필요한 기능을 찾기 위한 키워드 검색 기능을 추가하였다.

기능이 고도화 됨에 따라, 만두는 템플릿 관리 서비스를 다른 사람들에게도 제공하고 싶어졌다.
각자 개인 템플릿을 관리하는 private 한 공간을 제공하기 위해 로그인 기능이 추가되었다

**Comment by kyum-q** [comment]
# 3. 태그

**Reply by HoeSeong123** [reply]
- 카테고리와 비슷한 이유
- 검색의 범위를 조금이라도 더 넓히고 싶음

**Reply by kyum-q** [reply]
## 태그 (한 글에 여러개가 있을 수 있음)
템플릿의 메타데이터를 태그로 표시하자
각 태그를 클릭하면 그에 해당하는 글이 보이도록 하자

### 장점
- 카테고리는 커스텀하는 폴더명 느낌이다
- 메타 데이터 느낌으로 사용

**Reply by zangsu** [reply]
코드 템플릿에 세부 사항이 되는 정보들을 표시해 줄 수 있는 가장 좋은 방법이 태그임

> ex) 
제목 : Cookie 를 사용하는 로그인
description : Spring boot 3 에서 Cookie 를 사용해 인증/인가 정보를 제공한다.
category : Spring
tag : `http-only`, `https`, `same-site`

**Reply by kyum-q** [reply]
## 시나리오:

- 만두는 우테코 코딩테스트를 준비하기 위해 '코드잽'을 실행합니다.
- 만두는 몇 번째 프리코스에서 구현했는지에 따라 카테고리를 구분하였습니다.
   - 카테고리 [로또, 자동차경주, 크리스마스] 
- 만두는 우테코 동안 구현한 코드 템플릿을 작성합니다.
- 만두는 '자동차 경주' 미션 때 구현한 'JWT를 이용한 로그인' 코드 스니펫을 신속하게 찾고자 합니다.
- 하지만 열정적인 만두는 너무 많은 템플릿을 만들어놔서 찾기 어려웠습니다.
- 만두는 '자동차 경주' 카테고리에서 '로그인'으로 검색을 합니다.
- 하지만 만두는 템플릿의 이름을 'JWT이용한 Login 구현'으로 생성해놔서 검색 결과로 JWT가 없었습니다.
- 그리고 포괄적인 이름의 템플릿인 경우 JWT를 이용했는지 템플릿 상세 페이지를 확인해야 했습니다.
- 만두는 검색을 편리하게 하기 위해 템플릿에 'JWT', 'Login' 태그를 추가했습니다.
- 만두는 시간이 지나 우테코 최종 코딩 테스트를 보러왔습니다.
- 코딩 테스트에서 JWT를 이용한 로그인 기능 구현이 필수 구현 항목에 있었습니다.
- 만두는 '코드잽'을 실행합니다.
- 만두는 '자동차 경주' 카테고리로 들어갑니다. 
- 태그 목록에서 'JWT', 'Login'를 선택해 JWT 로그인과 관련된 코드템플릿만 확인합니다.
- 그리고 태그 목록과 제목 등을 통해 JWT의 로그인 기능을 빠르게 찾습니다.
- 만두는 한번의 클릭으로 쉽게 코드를 재사용합니다.
- 만두는 원하는 기능의 코드를 빠르게 찾아 시간을 대폭 단축하여 우테코 최종 코딩테스트에 합격합니다.

**Comment by kyum-q** [comment]
## 시나리오

**Reply by kyum-q** [reply]
## 3차 스프린트 시나리오:

- 만두는 우테코 코딩테스트를 준비하기 위해 '코드잽'을 실행합니다.
- 만두는 몇 번째 프리코스에서 구현했는지에 따라 카테고리를 구분하였습니다.
   - 카테고리 [로또, 자동차경주, 크리스마스] 
- 만두는 우테코 동안 구현한 코드 템플릿을 작성합니다.
- 만두는 '자동차 경주' 미션 때 구현한 'JWT를 이용한 로그인' 코드 스니펫을 신속하게 찾고자 합니다.
- 하지만 열정적인 만두는 너무 많은 템플릿을 만들어놔서 찾기 어려웠습니다.
- 만두는 '자동차 경주' 카테고리에서 '로그인'으로 검색을 합니다.
- 하지만 만두는 템플릿의 이름을 'JWT이용한 Login 구현'으로 생성해놔서 검색 결과로 JWT가 없었습니다.
- 그리고 포괄적인 이름의 템플릿인 경우 JWT를 이용했는지 템플릿 상세 페이지를 확인해야 했습니다.
- 만두는 검색을 편리하게 하기 위해 템플릿에 'JWT', 'Login' 태그를 추가했습니다.
- 만두는 시간이 지나 우테코 최종 코딩 테스트를 보러왔습니다.
- 코딩 테스트에서 JWT를 이용한 로그인 기능 구현이 필수 구현 항목에 있었습니다.
- 만두는 '코드잽'을 실행합니다.
- 만두는 '자동차 경주' 카테고리로 들어갑니다. 
- 태그 목록에서 'JWT', 'Login'를 선택해 JWT 로그인과 관련된 코드템플릿만 확인합니다.
- 그리고 태그 목록과 제목 등을 통해 JWT의 로그인 기능을 빠르게 찾습니다.
- 만두는 한번의 클릭으로 쉽게 코드를 재사용합니다.
- 만두는 원하는 기능의 코드를 빠르게 찾아 시간을 대폭 단축하여 우테코 최종 코딩테스트에 합격합니다.
